File: backend-python/routes/vendors.py
from typing import Optional

from fastapi import APIRouter

from config.database import get_db, is_connected
from utils.helpers import serialize_list

router = APIRouter()


# GET /api/vendors
@router.get("/")
async def list_vendors(
    category: Optional[str] = None,
    verified: Optional[str] = None,
    search: Optional[str] = None,
):
    if not is_connected():
        return {"success": False, "message": "Database not connected"}, 503

    db = get_db()
    query: dict = {}
    if category:
        query["category"] = category
    if verified is not None:
        query["verified"] = verified.lower() == "true"
    if search:
        query["$or"] = [
            {"name": {"$regex": search, "$options": "i"}},
            {"location": {"$regex": search, "$options": "i"}},
        ]

    cursor = db.vendors.find(query).sort("rating", -1)
    vendors = serialize_list(await cursor.to_list(length=100))
    return {"success": True, "vendors": vendors}

File: backend-node/routes/vendors.js
// ============================================================
// routes/vendors.js — Vendor listing
// ============================================================

const express  = require('express');
const mongoose = require('mongoose');
const vendorRouter = express.Router();

let Vendor;
try {
  Vendor = require('../models/Vendor');
} catch {
  Vendor = null;
}

const dbReady = () => mongoose.connection.readyState === 1;

// GET /api/vendors
vendorRouter.get('/', async (req, res, next) => {
  try {
    if (Vendor && dbReady()) {
      const { category, verified, search } = req.query;
      const query = {};
      if (category) query.category = category;
      if (verified)  query.verified = verified === 'true';
      if (search)    query.$or = [{ name: new RegExp(search, 'i') }, { location: new RegExp(search, 'i') }];
      const vendors = await Vendor.find(query).sort({ rating: -1 });
      return res.json({ success: true, vendors });
    }

    // ── No DB: return 503 so frontend falls back to local Data.vendors ──
    res.status(503).json({ success: false, message: 'Database not connected' });
  } catch (err) {
    next(err);
  }
});

module.exports = vendorRouter;
